File: Frontend/src/app/product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Feature, Product } from './Product';

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  
  private apiUrl = 'http://localhost:8080/api/products';

  constructor(private http: HttpClient) {}
  internalName: string= "";

  getAllProducts(): Observable<Product[]> {
    return this.http.get<Product[]>(this.apiUrl);
  }

  getProductByInternalName(internalName: string): Observable<Product> {
    const url = `${this.apiUrl}/${internalName}`;
    return this.http.get<Product>(url);
  }

  getFeaturesByProductInternalName(productInternalName: string): Observable<Feature[]> {
    const url = `${this.apiUrl}/${productInternalName}/features`;
    return this.http.get<Feature[]>(url);
  }

  addProduct(product: Product): Observable<void> {
    const url = `${this.apiUrl}/add`;
    return this.http.post<void>(url, product);
  }

  updateProduct(product: Product): Observable<Product> {
    const url = `${this.apiUrl}/update`;
    return this.http.put<Product>(url, product);
  }

  deleteProduct(product: Product) {

    const url = `${this.apiUrl}/delete`;
    console.log(url);
    
    return this.http.post(url,product.internalName);
  }

  saveProductName(internalName: string): void{
    this.internalName = internalName;
  }

  getProductName(): string{
    return this.internalName;
  }
  
  }

File: src/main/java/com/learnings/capstone/service/ProductService.java
package com.learnings.capstone.service;

import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.learnings.capstone.dto.FeatureDTO;
import com.learnings.capstone.dto.ProductDTO;
import com.learnings.capstone.entity.Feature;
import com.learnings.capstone.entity.Product;
import com.learnings.capstone.exception.CatalogBusinessException;
import com.learnings.capstone.repository.ProductRepository;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ProductService {

    @Autowired 
    private ProductRepository productRepository;

    public void addProduct(ProductDTO productDTO) {
        Product product = convertDTOToEntity(productDTO);
        productRepository.save(product);
    }

    public ProductDTO getProductById(Long productId) {
        Product product = getProductEntityById(productId);
        return convertEntityToDTO(product);
    }

    public ProductDTO getProductByInternalName(String productInternalName ) {
        Product product = getProductEntityByInternalName(productInternalName);
        return convertEntityToDTO(product);
    }

    public List<ProductDTO> getAllProducts() {
        List<Product> products = productRepository.findAll();
        return products.stream()
                .map(this::convertEntityToDTO)
                .collect(Collectors.toList());
    }

    public ProductDTO updateProduct(ProductDTO productDTO) {
        Product existingProduct = productRepository.getProductEntityByInternalName(productDTO.getInternalName());
    
        if (existingProduct == null) {
            // Handle the case where the product is not found
            throw new CatalogBusinessException("Product not found with internal name: " + productDTO.getInternalName());
        }
    
        updateProductFromDTO(existingProduct, productDTO);
        productRepository.save(existingProduct);
    
        return convertEntityToDTO(existingProduct);
    }
    

    public void deleteProduct(String productInternalName) {
        Product existingProduct = productRepository.getProductEntityByInternalName(productInternalName);
        productRepository.delete(existingProduct);
    }

    private Product convertDTOToEntity(ProductDTO productDTO) {
        Product product = new Product();
        BeanUtils.copyProperties(productDTO, product);
        // Set other properties as needed
        return product;
    }

    private ProductDTO convertEntityToDTO(Product product) {
        ProductDTO productDTO = new ProductDTO();
        BeanUtils.copyProperties(product, productDTO);
        // Set other properties as needed
        return productDTO;
    }

    private void updateProductFromDTO(Product product, ProductDTO productDTO) {
        BeanUtils.copyProperties(productDTO, product);
        // Update other properties as needed
    }

    public Product getProductEntityById(Long productId) {
        return productRepository.findById(productId)
                .orElseThrow(() -> new CatalogBusinessException("Product not found with id " + productId));
    }

    public List<FeatureDTO> getFeaturesByProductInternalName(String internalName) {
        Product product = getProductEntityByInternalName(internalName);
    
        if (product != null && product.getFeatures() != null) {
            List<FeatureDTO> featureDTOs = product.getFeatures().stream()
                    .map(this::convertFeatureEntityToDTO)
                    .collect(Collectors.toList());
    
            // Log the features for debugging
            System.out.println("Features for product " + internalName + ": " + featureDTOs);
    
            return featureDTOs;
        } else {
            // Handle the case where features are null or the product is not found
            return Collections.emptyList(); // Or throw an exception, return a default list, etc.
        }
    }
    
    

    public Product getProductEntityByInternalName(String internalName) {
        return productRepository.getProductEntityByInternalName(internalName);
    }

    public FeatureDTO convertFeatureEntityToDTO(Feature feature) {
        FeatureDTO featureDTO = new FeatureDTO();
        BeanUtils.copyProperties(feature, featureDTO);
        // Set other properties as needed
        return featureDTO;
    }
}
